internal/cuckoo/cuckoo.go

```diff
@@ -0,0 +1,239 @@
+// so the index of the Cuckoo.items slice is +1 compared to the index
+// of the input slice you use.
+type Cuckoo struct {
+	items        [][]byte
+	hashIndices  []byte
+	bucketLookup []uint64
+	// Total bucket count, len(bucket)
+	bucketSize uint64
+	// 3 hash functions h_0, h_1, h_2
+	hashers [Nhash]hash.Hasher
+}
+
+// NewCuckoo instantiates a Cuckoo struct with a bucket of size Factor * size,
+// a stash and 3 seeded hash functions for the 3-way cuckoo hashing.
+func NewCuckoo(size uint64, seeds [Nhash][]byte) *Cuckoo {
+	bSize := max(1, uint64(Factor*float64(size)))
+	var hashers [Nhash]hash.Hasher
+	for i, s := range seeds {
+		hashers[i], _ = hash.New(HashFunc, s)
+	}
+
+	return &Cuckoo{
+		// extra element is "keeper" to which the bucketLookup can be directed
+		// when there is no element present in the bucket.
+		items:        make([][]byte, size+1),
+		hashIndices:  make([]byte, size+1),
+		bucketLookup: make([]uint64, bSize),
+		bucketSize:   bSize,
+		hashers:      hashers,
+	}
+}
+
+// NewDummyCuckoo instantiates a Cuckoo struct that does not allocate buckets.
+func NewDummyCuckoo(size uint64, seeds [Nhash][]byte) *Cuckoo {
+	bSize := max(1, uint64(Factor*float64(size)))
+	var hashers [Nhash]hash.Hasher
+	for i, s := range seeds {
+		hashers[i], _ = hash.New(HashFunc, s)
+	}
+
+	return &Cuckoo{
+		bucketSize: bSize,
+		hashers:    hashers,
+	}
+}
+
+// GetHasher returns the first seeded hash function from a cuckoo struct.
+func (c *Cuckoo) GetHasher() hash.Hasher {
+	return c.hashers[0]
+}
+
+// BucketIndices returns the 3 possible bucket indices of an item
+func (c *Cuckoo) BucketIndices(item []byte) (idxs [Nhash]uint64) {
+	for i := range idxs {
+		idxs[i] = c.hashers[i].Hash64(item) % c.bucketSize
+	}
+
+	return idxs
+}
+
+// GetBucket returns the index in a given bucket which represents the value in
+// the list of identifiers to which it points.
+func (c *Cuckoo) GetBucket(bIdx uint64) (uint64, error) {
+	if bIdx > c.bucketSize {
+		return 0, fmt.Errorf("failed to retrieve item in bucket #%v", bIdx)
+	}
+	return c.bucketLookup[bIdx], nil
+}
+
+// GetItemWithHash returns the item at a given index along with its
+// hash index.
+func (c *Cuckoo) GetItemWithHash(idx uint64) (item []byte, hIdx uint8) {
+	if idx > uint64(len(c.items)-1) {
+		return nil, 0
+	}
+
+	return c.items[idx], c.hashIndices[idx]
+}
+
+// Exists returns true if an item is inserted in cuckoo, false otherwise
+func (c *Cuckoo) Exists(item []byte) (bool, byte) {
+	bucketIndices := c.BucketIndices(item)
+	for hIdx, bIdx := range bucketIndices {
+		if bytes.Equal(c.items[c.bucketLookup[bIdx]], item) {
+			return true, byte(hIdx)
+		}
+	}
+	return false, 0
+}
+
+// Insert tries to insert a given item (at index, idx) to the bucket
+// in available slots, otherwise, it evicts a random occupied slot,
+// and reinserts evicted item.
+// Returns an error msg if all failed.
+func (c *Cuckoo) Insert(input <-chan []byte) error {
```

Comment:
Using a channel should be left to the caller of the cuckoo package. We are allocating a cuckoo hash that can contain N elements. The cuckoo hash should keep track of how many elements have been inserted, only insert in the next available spot and error out when there's no more space left. Usually we would do that with something like a CuckooBuilder which has a freeze method and would return the actual cuckoo with only what's needed on the read side but I think it's okay to do it directly in the Cuckoo struct.